### DISCUSSION TITLE
RFC: Static Path Props

### DISCUSSION THREAD

**Discussion by kristojorg** [vercel/next.js#18052] [Ideas]
# Description 
This is a proposal to solve two concurrent problems:

1. App wide static props: You have some data you need on every page of the app, and you don't want to refetch it when statically rendering each page.
1. Path wide static props: You have some data you need on every sub-route of a path, and you don't want to refetch it when statically rendering every page in that path. Eg: You have `/blog/:blogSlug`, and every path there needs the list of all blog posts, but `/dashboard` doesn't need the same static data. 

To do this, I propose one change:

- `getStaticProps` can return a `pathProps` key, which will be available on all children of that path.

## Example

```js
// /pages/blog/index.js

function Blog({ posts }) {
  return (
    <ul>
      {posts.map((post) => (
        <li>{post.title}</li>
      ))}
    </ul>
  )
}

export async function getStaticProps() {
  const res = await fetch('https://.../posts')
  const posts = await res.json()

  return {
    props: {}, // you can still provide page-specific data here
    pathProps: { // this data will be passed to all sub-routes
      posts,
    },
  }
}

export default Blog
```
Then in your sub routes, it is also available without refetching:

```js
// pages/blog/[blogSlug].js

function Post({post, posts}) {
  return (
    <header>
      <ul>
        {posts.map((post) => (
          <li>{post.title}</li>
        ))}
      </ul>
    </header>
  );
}

export async function getStaticProps(ctx) {
  const post = await fetchPost(ctx.query.blogSlug)
  // ctx.pathProps.posts also holds the posts data fetched higher up the path hierarchy
  return {
    props: { post },
  }
}
```

### Notes

1. Any conflicts would be overwritten by children. We could alternatively have each path provide its props on a key identified by that path: `props.pathProps.posts` or similar, if desired.
2. If a path doesn't have any component at `/index.js`, it can still export a `getStaticProps` to fetch props for that route
3. The path props would revalidate on the same schedule as the other static props where they are fetched.
4. We could alternatively do this in a function other than `getStaticProps`, like `getStaticPathProps`.
5. To set app-wide data, you would use `pathProps` in `/pages/index.js`, not `_app.js`. This is because `/` is a parent to all other routes, so data returned in `pathProps` there would be passed down to every other route.

I'll explore what it would take to implement this change, but comments, questions, concerns, help is all welcome!

### Related Discussions

- https://github.com/vercel/next.js/discussions/10949
- https://github.com/vercel/next.js/discussions/11334
- https://github.com/vercel/next.js/discussions/17888
- https://github.com/vercel/next.js/discussions/17555

**Comment by Vadorequest** [comment]
How would you do that if you have a `/pages/[slug]/index` path, so that all paths under `/[slug]` would binherit from the data? Maybe something to do with getStaticPaths? 🤔 

But overall, the proposal looks great and would solve my current issue described at https://github.com/vercel/next.js/discussions/14949

**Reply by kristojorg** [reply]
`/pages/[slug]/index` would define a `getStaticProps` which would fetch the data necessary for that `slug` value, and then that would be run and cached whenever either the `/pages/[slug]/index` page is rendered or when a child route is rendered. All the data provided to `pathProps` in `/pages/[slug]/index` would also be available in `/pages/[slug]/child` too

**Comment by JonCognioDigital** [comment]
This is a fantastic idea. I am going to make a separate post about dealing with site-level or section-level data as I had a different solution but I will reference this one as it certainly seems to solve most of the problems I have and some I hadn't even thought of.

**Reply by JonCognioDigital** [reply]
Here is my soution to a slightly different (but related) problem: https://github.com/vercel/next.js/issues/20698

**Comment by justincy** [comment]
An alternative solution (which also solves many other problems) would be for Next.js to support nested layouts.

**Reply by kristojorg** [reply]
Is there an RFC or proposal/PR for this? I'm not sure exactly what you mean

**Reply by justincy** [reply]
No. The idea is just that layouts could be inherited based on path (so that only a portion of the page get re-rendered) and along with that you could also have them inherit props.

**Reply by JonCognioDigital** [reply]
And also that you could do data fetching in different parts of the heirarcy, which can currently only be done at page level.

**Reply by leerob** [reply]
Layout support will definitely come to Next.js, eventually. I agree layouts would be a better solution here.

**Comment by steven-tey** [comment]
Was this ever implemented? Would love to be able to solve the age-old app-wide/path-wide static props issue soon! 😄

**Comment by oranoran** [comment]
Awesome idea. I suggested something similar in #24737, using the fallback field.
